### ISSUE TITLE
LibJS: some operations return 0 instead of infinity

### ISSUE THREAD

**Issue by tim-st** [SerenityOS/serenity#5212]
```
> 2**30
1073741824
> 2**31
0
> 2**310
0
> 2**3100
Infinity
```

Only the first and the last result is expected, I think.

**Comment by linusg** [comment]
I don't think the exponentiation is actually broken, this seems to be an issue with the `double_to_string()` function that for some reason returns 0 for values >= 2^31 - apply this patch to confirm it's being called with the correct result:

```diff
--- a/Userland/Libraries/LibJS/Runtime/Value.cpp
+++ b/Userland/Libraries/LibJS/Runtime/Value.cpp
@@ -83,6 +83,7 @@ ALWAYS_INLINE bool both_bigint(const Value& lhs, const Value& rhs)
 
 static String double_to_string(double d)
 {
+    printf("Called double_to_string() with d = %f\n", d);
     // https://tc39.es/ecma262/#sec-numeric-types-number-tostring
     if (isnan(d))
         return "NaN";
```

The Lagom build is fine somehow:

| `js` on SerenityOS | `js` on Linux via Lagom |
| ------------------ | ----------------------- |
| ![image](https://user-images.githubusercontent.com/19366641/106506873-96234f80-64ca-11eb-9a8e-0de73bc53d6f.png) | ![image](https://user-images.githubusercontent.com/19366641/106506960-b3f0b480-64ca-11eb-8455-7751861269fb.png) |

**Comment by linusg** [comment]
...which lets me think this is probably and issue with using `size_t` :thinking:

**Comment by tim-st** [comment]
also `(2**31) + 1` prints `0`, so somehow the `1` is not added on the (incorrect) 0.

**Comment by IdanHo** [comment]
This likely due to the fact that 2147483648 (2**31) doesnt fit in an int, so when modf from LibMath (which is used to seperate the double to its integer and fractional parts in double_to_string) casts the double to an int
```c++
*intpart = (double)((int)(x));
```
it results in undefined behaviour (https://eel.is/c++draft/conv#fpint-1) which in our case is a negative value, which immediately break the while loop in double_to_string which converts the integer part to a string
```c++
while (intPart > 0) {
```

**Comment by IdanHo** [comment]
Fixed by 6622ad88952cf27f432cd4bf51b1e815afcd986e - can be closed
